crates/red_knot_python_semantic/src/types.rs

```diff
@@ -296,6 +297,42 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Return true if this type is assignable to type `other`.
+    #[allow(unused)]
+    pub(crate) fn is_assignable_to(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        if self.is_equivalent_to(db, other) {
+            return true;
+        }
+        match (self, other) {
+            (Type::Unknown | Type::Any | Type::Never, _) => true,
+            (_, Type::Unknown | Type::Any) => true,
+            (Type::IntLiteral(_), Type::Instance(class)) if class.is_builtin_named(db, "int") => {
+                true
+            }
+            (Type::StringLiteral(_), Type::LiteralString) => true,
+            (Type::StringLiteral(_) | Type::LiteralString, Type::Instance(class))
+                if class.is_builtin_named(db, "str") =>
+            {
+                true
+            }
+            (Type::BytesLiteral(_), Type::Instance(class))
+                if class.is_builtin_named(db, "bytes") =>
+            {
+                true
+            }
+            // TODO
+            _ => false,
```

Comment:
nit: I know it can be tedious to list out all the missing branches here, but it means that it's guaranteed that we won't forget to update the method when we add new `Type` variants, so I'd prefer it

Reply:
I don't think this idea works when we're talking about a double-match like this one. I definitely don't think we want to list out every possible pair of Type variants; we'd end up with a match statement of n^2 size, with most arms just returning false.

The inherent nature of assignability is that the default case for two unrelated type variants is "not assignable". I think it makes sense for the match statement to reflect this, and handle only the cases where they are or might be assignable.

Reply:
Okay, fair enough!